docs/topics/db/instrumentation.txt

```diff
@@ -0,0 +1,121 @@
+code, and so do not need a separate factory like middleware do.
+
+Installing a wrapper is done in a context manager -- so the wrappers are
+temporary and can be made specific to some flow in your code.
+
+Probably the most trivial example of a wrapper is an execution blocker as
+mentioned above. It could look like this::
+
+    def blocker(*args):
+        raise Exception("No database access allowed here")
+
+And it would be used in a view to block queries from the template like so::
+
+    from django.db import connection
+    from django.shortcuts import render
+
+    def my_view(request):
+        context = ...  # Code to generate context with all data
+        template_name = ...
+        with connection.execute_wrapper(blocker):
+            return render(request, template_name, context)
+
+The parameters sent to the wrappers are:
+
+* ``execute`` -- a callable, which should be invoked with the rest of the
+  parameters in order to execute the query.
+
+* ``sql`` -- a ``str``, the SQL text to be sent to the database engine.
+
+* ``params`` -- a list/tuple of parameter values for the SQL command, or a
+  list/tuple of lists/tuples if the wrapped call is ``executemany()``
+
+* ``many`` -- a ``bool`` indicating whether the ultimately invoked call is
+  ``execute()`` or ``executemany()`` (and whether ``params`` is expected to be
+  a sequence of values, or a sequence of sequences of values).
+
+* ``context`` -- a dictionary with further data about the context of
+  invocation. For now, this includes the connection and cursor.
+
+Using them, a slightly more complex version of the blocker could include the
+connection name in the error message::
+
+    def blocker(execute, sql, params, many, context):
+        alias = context['connection'].alias
+        raise Exception("Access to database '{}' blocked here".format(alias))
+
+For a more complete example, a query logger could look like this::
+
+    import time
+
+    class QueryLogger:
+
+        def __init__(self):
+            self.queries = []
+
+        def __call__(execute, sql, params, many, context):
+            current_query = {'sql': sql, 'params': params, 'many': many}
+            start = time.time()
+            try:
+                result = execute(sql, params, many, context)
+            except Exception as e:
+                current_query['status'] = 'error'
+                current_query['exception'] = e
+                raise
+            else:
+                current_query['status'] = 'ok'
+                return result
+            finally:
+                duration = time.time() - start
+                current_query['duration'] = duration
+                self.queries.append(current_query)
+
+To use this effectively, you would create a logger object and install it as a
+wrapper::
+
+    from django.db import connection
+
+    ql = QueryLogger()
+    with connection.execute_wrapper(ql):
+        # Do some interesting querying
+    # Access ql.queries here
```

Comment:
you could make this valid python by calling fake functions instead like `do_some_queries()` and `print(ql.queries)`